File: 922-PossibleBipartition/922-PossibleBipartition.java
// Last updated: 12/6/2025, 5:45:41 am
class Solution {
  public boolean possibleBipartition(int n, int[][] dislikes) {    
    Map<Integer, Set<Integer>> graph = buildGraph(dislikes, n);
    UnionFind uf = new UnionFind(n);
    
    for(int i = 1; i < n+1; i++) {
      Set<Integer> neighbors = graph.get(i);
      if(neighbors.size() == 0) continue;
      int firstNeighbor = neighbors.iterator().next(); // gets the first item in the set
      
      for(int neighbor : neighbors) {
        if(uf.isConnected(i, neighbor)) return false; // if vertex i is connected with any of its neighbors, graph is not bipartite
        uf.union(firstNeighbor, neighbor); // unionize all its neighbors
      }
    }
        
    return true;
  }
  
  public Map<Integer, Set<Integer>> buildGraph(int[][] edges, int n) {
    Map<Integer, Set<Integer>> graph = new HashMap<>();
    
    for(int i = 1; i < n+1; i++) graph.put(i, new HashSet<>());
    
    for(int[] edge : edges) {
      int src = edge[0];
      int dest = edge[1];
      
      graph.get(src).add(dest);
      graph.get(dest).add(src);
    }
    
    return graph;
  }
  
  class UnionFind {
    int[] parent;
    int[] rank;
    
    public UnionFind(int n) {
      parent = new int[n+1];
      rank = new int[n+1];
      
      for(int i = 0; i < n+1; i++) {
        parent[i] = i;
      }
    }
    
    public void union(int x, int y) {
      int rootX = find(x);
      int rootY = find(y);
      
      if(rootX != rootY) {
        if(rank[rootX] > rank[rootY]) {
          rank[rootX]++;
          parent[rootY] = rootX;
        } else {
          rank[rootY]++;
          parent[rootX] = rootY;
        }
      }
    }
    
    public int find(int x) {
      if(parent[x] == x) return x;
      
      int root = find(parent[x]);
      parent[x] = root; //path compression
      return root;
    }
    
    public boolean isConnected(int x, int y) {
      return find(x) == find(y);
    }
  }
}

File: 922-PossibleBipartition/922-PossibleBipartition.py
# Last updated: 12/6/2025, 5:45:36 am
class Union:
    def __init__(self, n):
        self.parents = []
        self.size = []
        for i in range(n):
            self.parents.append(i)
            self.size.append(0)
        
    def find(self, node):
        if node != self.parents[node]:
            self.parents[node] = self.find(self.parents[node])
        return self.parents[node]
    
    def union(self, node1, node2):
        root1 = self.find(node1)
        root2 = self.find(node2)

        if root1 == root2: return 0

        if self.size[root1] < self.size[root2]:
            root2, root1 = root1, root2
        
        self.size[root1] += 1
        self.parents[root2] = root1
        return 1
    
    def isConnected(self, x, y):
        return self.find(x) == self.find(y)

class Solution:
    def possibleBipartition(self, n: int, dislikes: List[List[int]]) -> bool:
        graph = defaultdict(list)
        for d in dislikes:
            graph[d[0]-1].append(d[1]-1)
            graph[d[1]-1].append(d[0]-1)

        uf = Union(n)

        for i in range(n):
            for v in graph[i]:
                if uf.isConnected(i, v): return False
                uf.union(graph[i][0], v)
        return True

        